src/powershell/Program.cs

```diff
@@ -19,7 +63,433 @@ public sealed class ManagedPSEntry
+            int j = 1;
+            for (; i < arg.Length; i++, j++)
+            {
+                char c = arg[i];
+
+                if (c == '\'')
+                {
+                    // /bin/sh quote escaping uses backslashes
+                    span[j] = '\\';
+                    j++;
+                }
+
+                span[j] = c;
+            }
+
+            span[j] = '\'';
+        }
+
+        /// <summary>
+        /// If the given exit code is negative, throws a StartupException.
+        /// </summary>
+        /// <param name="call">The native call that was attempted.</param>
+        /// <param name="code">The exit code it returned.</param>
+        private static void ThrowOnFailure(string call, int code)
+        {
+            if (code < 0)
+            {
+                code = Marshal.GetLastWin32Error();
+                Console.Error.WriteLine($"Call to '{call}' failed with errno {code}");
+                throw new StartupException(call, code);
+            }
+        }
+
+        /// <summary>
+        /// The `execv` POSIX syscall we use to exec /bin/sh.
+        /// </summary>
+        /// <param name="path">The path to the executable to exec.</param>
+        /// <param name="args">
+        /// The arguments to send through to the executable.
+        /// Array must have its final element be null.
+        /// </param>
+        /// <returns>
+        /// An exit code if exec failed, but if successful the calling process will be overwritten.
+        /// </returns>
+        [DllImport("libc",
+            EntryPoint = "execv",
+            CallingConvention = CallingConvention.Cdecl,
+            CharSet = CharSet.Ansi,
+            SetLastError = true)]
+        private static extern int Exec(string path, string[] args);
+
+        /// <summary>
+        /// The `readlink` POSIX syscall we use to read the symlink from /proc/self/exe
+        /// to get the executable path of pwsh on Linux.
+        /// </summary>
+        /// <param name="pathname">The path to the symlink to read.</param>
+        /// <param name="buf">Pointer to a buffer to fill with the result.</param>
+        /// <param name="size">The size of the buffer we have supplied.</param>
+        /// <returns>The number of bytes placed in the buffer.</returns>
+        [DllImport("libc",
+            EntryPoint = "readlink",
+            CallingConvention = CallingConvention.Cdecl,
+            CharSet = CharSet.Ansi,
+            SetLastError = true)]
+        private static extern IntPtr ReadLink(string pathname, IntPtr buf, UIntPtr size);
+
+        /// <summary>
+        /// The `getpid` POSIX syscall we use to quickly get the current process PID on macOS.
+        /// </summary>
+        /// <returns>The pid of the current process.</returns>
+        [DllImport("libc",
+            EntryPoint = "getpid",
+            CallingConvention = CallingConvention.Cdecl,
+            CharSet = CharSet.Ansi,
+            SetLastError = true)]
```

Comment:
We can remove the last two unused arguments.

Reply:
`CharSet.Ansi` maybe be default on Unix platform, which I don't know for sure. But I think `SetLasterror = true` is needed if you are calling `GetLastWin32Error()` after P/Invoke.